File: CppPractice/CppLeetcode/Leetcode1367.cpp
#include<deque>
using namespace std;

class Solution {

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

public:
    bool dfs(ListNode* head, TreeNode *node)
    {
        if (head==nullptr)
        {
            return true;
        }
        if (node==nullptr)
        {
            return false;
        }
        if (head->val != node->val)
        {
            return false;
        }
        return dfs(head->next,node->left) || dfs(head->next,node->right);
    }
    bool isSubPath(ListNode* head, TreeNode* root) 
    {
        deque<TreeNode*> theQueue;
        theQueue.push_back(root);
        theQueue.pop_front();
        while (theQueue.size()>0)
        {
            TreeNode *node = theQueue.front();
            theQueue.pop_front();
            if (dfs(head,node))
            {
                return true;
            }
            else
            {
                if (node->left)
                {
                    theQueue.push_back(node->left);
                }
                if (node->right)
                {
                    theQueue.push_back(node->right);
                }
                
                
            }
            

        }
        return false;
        
    }
};

File: CSharpPractice/LeetcodePractice/Leetcode1367.cs
// Definition for singly-linked list.
public class Solution1367 {
    private bool dfs(ListNode listNode, TreeNode treeNode)
    {
        if (listNode==null)
        {
            return true;
        }
        if (treeNode==null)
        {
            return false;
        }
        if (listNode.val!=treeNode.val)
        {
            return false;
            
        }
        return dfs(listNode.next,treeNode.right)||dfs(listNode.next,treeNode.left);
    }
    public bool IsSubPath(ListNode head, TreeNode root) 
    {
        LinkedList<TreeNode> theDeque = new LinkedList<TreeNode>();
        theDeque.AddLast(root);

        while (theDeque.Count>0)
        {
            TreeNode node = theDeque.First.Value;
            theDeque.RemoveFirst();
            if (dfs(head,node))
            {
                return true;
                
            }
            if (node.right!=null)
            {
                theDeque.AddLast(node.right);
            }
            if (node.left!=null)
            {
                theDeque.AddLast(node.left);
            }
        }

        return false;
    }
}